src/random_ray/random_ray.cpp

```diff
@@ -0,0 +1,288 @@
+#include "openmc/random_ray/random_ray.h"
+
+#include "openmc/geometry.h"
+#include "openmc/message_passing.h"
+#include "openmc/mgxs_interface.h"
+#include "openmc/random_ray/flat_source_domain.h"
+#include "openmc/search.h"
+#include "openmc/settings.h"
+#include "openmc/simulation.h"
+#include "openmc/source.h"
+
+namespace openmc {
+
+//==============================================================================
+// Non-method functions
+//==============================================================================
+
+// returns 1 - exp(-tau)
+// Equivalent to -(_expm1f(-tau)), but faster
+// Written by Colin Josey.
+inline float cjosey_exponential(const float tau)
+{
+  const float c1n = -1.0000013559236386308f;
+  const float c2n = 0.23151368626911062025f;
+  const float c3n = -0.061481916409314966140f;
+  const float c4n = 0.0098619906458127653020f;
+  const float c5n = -0.0012629460503540849940f;
+  const float c6n = 0.00010360973791574984608f;
+  const float c7n = -0.000013276571933735820960f;
+
+  const float c0d = 1.0f;
+  const float c1d = -0.73151337729389001396f;
+  const float c2d = 0.26058381273536471371f;
+  const float c3d = -0.059892419041316836940f;
+  const float c4d = 0.0099070188241094279067f;
+  const float c5d = -0.0012623388962473160860f;
+  const float c6d = 0.00010361277635498731388f;
+  const float c7d = -0.000013276569500666698498f;
+
+  float x = -tau;
+  float num, den;
+
+  den = c7d;
+  den = den * x + c6d;
+  den = den * x + c5d;
+  den = den * x + c4d;
+  den = den * x + c3d;
+  den = den * x + c2d;
+  den = den * x + c1d;
+  den = den * x + c0d;
+
+  num = c7n;
+  num = num * x + c6n;
+  num = num * x + c5n;
+  num = num * x + c4n;
+  num = num * x + c3n;
+  num = num * x + c2n;
+  num = num * x + c1n;
+  num = num * x;
+
+  const float exponential = num / den;
+  return exponential;
+}
+
+//==============================================================================
+// RandomRay implementation
+//==============================================================================
+
+// Static Variable Declarations
+double RandomRay::distance_inactive_;
+double RandomRay::distance_active_;
+unique_ptr<Source> RandomRay::ray_source_;
+
+RandomRay::RandomRay()
+  : negroups_(data::mg.num_energy_groups_),
+    angular_flux_(data::mg.num_energy_groups_),
+    delta_psi_(data::mg.num_energy_groups_)
+{}
+
+RandomRay::RandomRay(uint64_t ray_id, FlatSourceDomain* domain) : RandomRay()
+{
+  initialize_ray(ray_id, domain);
```

Comment:
I thought about that as well, but there are a few potential use cases where we might want to re-initialize a ray without having to spend the time free'ing and re-allocating memory for it. In particular, it would be nice to run the initialization kernel on GPU (as we do with OpenMC OMP offloading branch), rather than having to run it on CPU. Could also be nice as a future optimization to have the parallel region start outside the loop over rays, such that each thread can just allocate a single ray once and then just re-initialize it for each work item.